shared/django_apps/rollouts/models.py

```diff
@@ -0,0 +1,50 @@
+from random import choice
+
+from django.contrib.postgres import fields
+from django.db import models
+
+
+def default_random_salt():
+    chars = []
+    ALPHABET = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+    for _ in range(16):
+        chars.append(choice(ALPHABET))
+    return "".join(chars)
+
+
+class FeatureFlag(models.Model):
+    name = models.CharField(max_length=200, primary_key=True)
+    proportion = models.DecimalField(default=0, decimal_places=3, max_digits=3)
+    salt = models.CharField(max_length=10000, default=default_random_salt)
+
+    class Meta:
+        db_table = "feature_flags"
+        constraints = [
+            models.UniqueConstraint(fields=["name"], name="feature_flag_name")
+        ]
+        # TODO: assert that variant proportions sum to 1
+
+
+class FeatureFlagVariant(models.Model):
+    name = models.CharField(max_length=200, primary_key=True)
+    feature_flag = models.ForeignKey(
+        "FeatureFlag", on_delete=models.CASCADE, related_name="variants"
+    )
+    proportion = models.DecimalField(default=0, decimal_places=3, max_digits=3)
+    enabled = models.BooleanField(default=False)
```

Comment:
I find a boolean is slightly awkward for this. What happens when we have a feature with three different variants?

Reply:
What are other variants you could think `enabled` could have? Is there a difference bw `enabled` and `state` or something like that?

Reply:
This is a bit of a contrived example but say we had a feature like: 

`NewRateLimiter` and we had three behaviours that leaked drops from a bucket at speeds `A`, `B`, `C`. We wouldn't be able to encode those three options with a single boolean. 

Especially with the current interface which expects a boolean for: is this feature on or off pretty much
```
if LIST_REPOS_GENERATOR_BY_OWNER_SLUG.check_value(user_id, default=False):
    new_behavior()
else:
    old_behavior()
```

I think this case of having more than 2 variants is rather rare though, so its probably fine to leave it as is?

Reply:
i was originally thinking the feature variant's actual value (`True`/`False`, or `A`/`B`/`C`) would still live in code, but that is also a little awkward

if we want to put the variant's value in the database, we could use a `JSONField`. if somebody enters `true` in the django admin UI it'll automatically be parsed into an actual boolean `True` in Python, or they could enter strings, arrays, objects, whatever